# Segmenting and Clustering Neighborhoods in Toronto
## Part Two

We will be using the BeautifulSoup package to scrape the table from the provided web address.

In [1]:
import pandas as pd
import requests
from bs4 import BeautifulSoup

In [2]:
website_url = requests.get('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').text

In [3]:
soup = BeautifulSoup(website_url, 'lxml')

In [4]:
table = soup.find_all('table', class_ = "wikitable sortable")

In [5]:
len(table)

1

Fortunately there is only one table on the page, meaning we do not have to narrow anything down to scrape our data.

In [6]:
table = table[0]
type(table)

bs4.element.Tag

Now it's time to convert the table to a dataframe.

In [7]:
table_rows = table.find_all('tr')

res = []
for tr in table_rows:
    td = tr.find_all('td')
    row = [tr.text.strip() for tr in td if tr.text.strip()]
    if row:
        res.append(row)
df = pd.DataFrame(res, columns=["Postcode", "Borough", "Neighbourhood"])
df=df.rename(columns = {'Postcode':'Postal Code', 'Neighbourhood':'Neighborhood'})
df.head(10)

Postal Code           Borough      Neighborhood
0         M1A      Not assigned      Not assigned
1         M2A      Not assigned      Not assigned
2         M3A        North York         Parkwoods
3         M4A        North York  Victoria Village
4         M5A  Downtown Toronto      Harbourfront
5         M5A  Downtown Toronto       Regent Park
6         M6A        North York  Lawrence Heights
7         M6A        North York    Lawrence Manor
8         M7A      Queen's Park      Not assigned
9         M8A      Not assigned      Not assigned

In [8]:
df.shape

(288, 3)

First we'll remove the postal codes that don't have a borough.

In [9]:
df1 = df[df.Borough != 'Not assigned']
df1.head(10)

Postal Code           Borough      Neighborhood
2          M3A        North York         Parkwoods
3          M4A        North York  Victoria Village
4          M5A  Downtown Toronto      Harbourfront
5          M5A  Downtown Toronto       Regent Park
6          M6A        North York  Lawrence Heights
7          M6A        North York    Lawrence Manor
8          M7A      Queen's Park      Not assigned
10         M9A         Etobicoke  Islington Avenue
11         M1B       Scarborough             Rouge
12         M1B       Scarborough           Malvern

In [10]:
df1.shape

(211, 3)

Then we'll assign the areas that don't have a neighborhood to their overarching borough.

In [11]:
df1.Neighborhood[df1.Neighborhood == 'Not assigned'] = df1.Borough

df1.head(10)

/opt/conda/envs/Python36/lib/python3.6/site-packages/pandas/core/generic.py:8682: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._update_inplace(new_data)
/opt/conda/envs/Python36/lib/python3.6/site-packages/IPython/core/interactiveshell.py:3267: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  exec(code_obj, self.user_global_ns, self.user_ns)


Postal Code           Borough      Neighborhood
2          M3A        North York         Parkwoods
3          M4A        North York  Victoria Village
4          M5A  Downtown Toronto      Harbourfront
5          M5A  Downtown Toronto       Regent Park
6          M6A        North York  Lawrence Heights
7          M6A        North York    Lawrence Manor
8          M7A      Queen's Park      Queen's Park
10         M9A         Etobicoke  Islington Avenue
11         M1B       Scarborough             Rouge
12         M1B       Scarborough           Malvern

Finally we'll group the neighborhoods by their postal code and borough.

In [12]:
df2 = df1.groupby(['Postal Code','Borough'])['Neighborhood'].apply(', '.join).reset_index()
df2.head()

Postal Code      Borough                            Neighborhood
0         M1B  Scarborough                          Rouge, Malvern
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2         M1E  Scarborough       Guildwood, Morningside, West Hill
3         M1G  Scarborough                                  Woburn
4         M1H  Scarborough                               Cedarbrae

In [13]:
df2.shape

(103, 3)

The dataframe has 103 rows and 3 columns. Now that the dataframe has been established and cleaned, we need to get the geospatial data for each postal code. We will be using the provided .csv file.

In [14]:
url="http://cocl.us/Geospatial_data"
df3=pd.read_csv(url)
df3.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Since the .csv file is easily read as a dataframe, all we have to do is merge it with our previous dataframe.

In [15]:
result = pd.merge(df2,df3, on='Postal Code')
result.head(20)

Postal Code      Borough  \
0          M1B  Scarborough   
1          M1C  Scarborough   
2          M1E  Scarborough   
3          M1G  Scarborough   
4          M1H  Scarborough   
5          M1J  Scarborough   
6          M1K  Scarborough   
7          M1L  Scarborough   
8          M1M  Scarborough   
9          M1N  Scarborough   
10         M1P  Scarborough   
11         M1R  Scarborough   
12         M1S  Scarborough   
13         M1T  Scarborough   
14         M1V  Scarborough   
15         M1W  Scarborough   
16         M1X  Scarborough   
17         M2H   North York   
18         M2J   North York   
19         M2K   North York   

                                         Neighborhood   Latitude  Longitude  
0                                      Rouge, Malvern  43.806686 -79.194353  
1              Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
2                   Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                              Woburn  43.770992 -79.216917  
4                                           Cedarbrae  43.773136 -79.239476  
5                                 Scarborough Village  43.744734 -79.239476  
6         East Birchmount Park, Ionview, Kennedy Park  43.727929 -79.262029  
7                     Clairlea, Golden Mile, Oakridge  43.711112 -79.284577  
8     Cliffcrest, Cliffside, Scarborough Village West  43.716316 -79.239476  
9                         Birch Cliff, Cliffside West  43.692657 -79.264848  
10  Dorset Park, Scarborough Town Centre, Wexford ...  43.757410 -79.273304  
11                                  Maryvale, Wexford  43.750072 -79.295849  
12                                          Agincourt  43.794200 -79.262029  
13            Clarks Corners, Sullivan, Tam O'Shanter  43.781638 -79.304302  
14  Agincourt North, L'Amoreaux East, Milliken, St...  43.815252 -79.284577  
15                                    L'Amoreaux West  43.799525 -79.318389  
16                                        Upper Rouge  43.836125 -79.205636  
17                                  Hillcrest Village  43.803762 -79.363452  
18                       Fairview, Henry Farm, Oriole  43.778517 -79.346556  
19                                    Bayview Village  43.786947 -79.385975